lib/Dialect/FIRRTL/Transforms/BlackBoxReader.cpp

```diff
@@ -0,0 +1,277 @@
+//===- BlackBoxReader.cpp - Move Verilog for black boxes --*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// Read Verilog source files for black boxes based on corresponding black box
+// annotations on the circuit and modules. Primarily based on:
+//
+// https://github.com/chipsalliance/firrtl/blob/master/src/main/scala/firrtl/transforms/BlackBoxSourceHelper.scala
+//
+//===----------------------------------------------------------------------===//
+
+#include "./PassDetails.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/Support/FileUtilities.h"
+
+#define DEBUG_TYPE "firrtl-blackbox-reader"
+
+using namespace circt;
+using namespace firrtl;
+
+/// Append a path to an existing path, replacing it if the other path is
+/// absolute. This mimicks the behaviour of `foo/bar` and `/foo/bar` being used
+/// in a working directory `/home`, resulting in `/home/foo/bar` and `/foo/bar`,
+/// respectively.
+static void appendPossiblyAbsolutePath(SmallVectorImpl<char> &base,
+                                       const Twine &suffix) {
+  if (llvm::sys::path::is_absolute(suffix)) {
+    base.clear();
+    suffix.toVector(base);
+  } else {
+    llvm::sys::path::append(base, suffix);
+  }
+}
+
+namespace {
+
+struct BlackBoxReaderPass : public BlackBoxReaderBase<BlackBoxReaderPass> {
+  BlackBoxReaderPass(llvm::Optional<StringRef> inputPrefix = {},
+                     llvm::Optional<StringRef> resourcePrefix = {}) {
+    if (inputPrefix.hasValue())
+      this->inputPrefix = inputPrefix.getValue().str();
+    if (resourcePrefix.hasValue())
+      this->resourcePrefix = resourcePrefix.getValue().str();
+  }
+
+  /// Emit the annotated source code for black boxes in a circuit.
+  void runOnOperation() override {
```

Comment:
Please move the bodies of these large methods out of line,  this makes it a bit easier to read the class declaration and reduces nesting of their bodies.